src/lib/operators/aggregate_hash.cpp

```diff
@@ -449,6 +509,14 @@ KeysPerChunk<AggregateKey> AggregateHash::_partition_by_groupby_keys() const {
                 ++chunk_offset;
               });
             }
+
+            auto previous_max = _expected_result_size.load();
+            while (previous_max < id_map.size()) {
```

Comment:
What does the loop actually do? I don't mean the max() part.

Reply:
I can't follow. The whole purpose of the loop is the max part.

Reply:
No, I meant a high level explanation of the whole loop (the compare_exchange part should be covered by reading the SO post).

Side note: a short comment why an atomic is needed for `_expected_result_size` in the hpp would help IMO.

Reply:
What is the "whole loop" for you? The for loop in line 420?

Regarding the second part, I added:

> _expected_result_size needs to be atomatically updated as the GROUP BY columns are processed in parallel.

Reply:
The loop in 516

Reply:
> The whole purpose of the loop is the max part.

Did you fully read the SO answer?

Reply:
Mmh, maybe that was a bit too snappish.

With `compare_exchange_strong`, we need to retry until the compare-and-swap has succeeded. That is the approach taken from the SO answer. I just slightly moved the code around because I do not like complex while-clauses with side-effects.

Reply:
That's not what bothers me. It's about what this loop actually does. Can you describe it?
I don't understand what you trying to achieve here.

Reply:
If we were alone (i.e., single-threaded), the code would look like this:

```
const auto previous_max = _expected_result_size;
if (previous_max < id_map.size()) {
  _expected_result_size = id_map.size();
}
```

However, we are not alone, so we need to atomically compare and swap the value:

```
auto previous_max = _expected_result_size;
if (previous_max < id_map.size()) {
  _expected_result_size.compare_exchange_strong(previous_max, id_map.size());
}
```

`compare_exchange_strong` checks if the value stored in `_expected_result_size` is `previous_max`. If it is, it updates it to `id_map.size()`. Ok - but what if it isn't?

In that case, it (1) updates `previous_max` to the value that it encounted instead and (2) returns false.

So if the first attempt fails, this would work on the second attempt:
```
auto previous_max = _expected_result_size;
if (previous_max < id_map.size()) {
  const auto success = _expected_result_size.compare_exchange_strong(previous_max, id_map.size());
  if (!success) {
    _expected_result_size.compare_exchange_strong(previous_max, id_map.size());
  }
}
```

So far so good, but what if the second attempt also fails?

This is how we end up with the while loop.

Reply:
Just saw your edit. Looks like we have been talking about different things. :/

> We will see at least `id_map.size()` different groups. We can use this knowledge to preallocate memory
for the results. Estimating the number of groups for multiple GROUP BY columns is somewhat hard, so we
simply take the number of groups created by the GROUP BY column with the highest number of distinct
values.

Reply:
I'd expect something like `std::max(id_map.size() * column_count, row_count())` to be a better estimate, but feel free to ignore this disussion. Besides that you can assume this issue to be resolved.

Reply:
Since we grow the results vector aggressively anyway, underestimating the size is not too much of an issue. With the current estimation, we are guaranteed to never overestimate. I believe this is more valuable, as we could otherwise see unexpected memory consumptions.

Reply:
Hmm, for a groupby of five columns, I would not expect it to be an overestimate to expect more keys than the max count of a single column. For a groupby column count of two, it would be a little bit bigger than the 1.5, but for larger groupby column counts I'd expect it to be a better estimate.

But anyhow  ... it's resolved.